### PR TITLE
chore: deprecate old config options

### PR DESCRIPTION
**Pull request by benlesh**
We need to remove these configuration options for v8, to help streamline the library and reduce size. Adding deprecations messages that give a little bit about why they are deprecated and some workaround information

Also adds some additional blurbs to documentation for promise-related methods on Observable.

Also adds some documentation to forEach with an example.

### CODE REVIEW

File: src/internal/config.ts [hunk 1]
```diff
@@ -18,6 +22,10 @@ export const config = {
    * where a multicast can be broken for all observers by a downstream consumer with
    * an unhandled error. DO NOT USE THIS FLAG UNLESS IT'S NEEDED TO BY TIME
    * FOR MIGRATION REASONS.
+   *
+   * @deprecated remove in v8. As of version 8, RxJS will no longer support synchronous throwing
```
Comment (kolodny) [inline src/internal/config.ts:26]: Good catch, I'm surprised this wasn't already "here be dragons"ed already

File: src/internal/Observable.ts [hunk 2]
```diff
@@ -251,12 +251,64 @@ export class Observable<T> implements Subscribable<T> {
   }
 
   /**
-   * @method forEach
+   * Used as a NON-CANCELLABLE means of subscribing to an observable, for use with
```
Comment (kolodny) [inline src/internal/Observable.ts:254]: Is it worth mentioning [rxjs-for-await](https://www.npmjs.com/package/rxjs-for-await)? I'm 50/50

File: src/internal/lastValueFrom.ts [hunk 3]
```diff
@@ -54,4 +60,4 @@ export function lastValueFrom<T>(source: Observable<T>) {
       },
     });
   });
-}
+}
```
Comment (kolodny) [inline src/internal/lastValueFrom.ts:63]: We should have an autoformatter that catches things like this

File: src/internal/Observable.ts (hunk 2 above)
Comment (benlesh) [reply src/internal/Observable.ts:254]: I'm not sure. Debatable. I'll leave it for now.

File: src/internal/firstValueFrom.ts [hunk 4]
```diff
@@ -13,6 +13,12 @@ import { Subscription } from './Subscription';
  * If the observable stream emits an error, the returned promise will reject
  * with that error.
  *
+ * **WARNING**: Only use this with observables you *know* will complete. If the source
```
Comment (jakovljevic-mladen) [inline src/internal/firstValueFrom.ts:16]: Hi there. Would this sentence be entirely true for `firstValueFrom`? There is a case where the source stream doesn't have to ever complete, it just have to emit at least one value. I would add ...`*know* will complete or emit at least once.`. Does this makes sense?

File: src/internal/Observable.ts [hunk 5]
```diff
@@ -251,12 +251,64 @@ export class Observable<T> implements Subscribable<T> {
...
+   * @param promiseCtor a constructor function used to instantiate the Promise
    * @return a promise that either resolves on observable completion or
    *  rejects with the handled error
+   * @deprecated remove in v8. Passing a Promise constructor will no longer be available
+   * in upcoming versions of RxJS. This is because it adds weight to the library, for very
+   * little benefit. If you need this functionality, it is recommended that you either
+   * polyfill Promise, or you create an adapter to convert the returned native promise
+   * to whatever promise implementation you wanted.
    */
+  forEach(next: (value: T) => void, promiseCtor: PromiseConstructorLike): Promise<void>;
```
Comment (cartant) [inline src/internal/Observable.ts:310]: Why is the `promiseCtor` no longer optional? Adding this signature means that the implementation signature won't be seen, so the fact that the parameter is option in the implementation signature would be meaningless.

Comment (cartant) [reply src/internal/Observable.ts:310]: Never mind. I've misread things.

**Pull request merged**